### PR TITLE
JSON internal API, IEEE754 base64/hex streaming, weights endpoint optimization

### PR DESCRIPTION
**Pull request by ktsaou**
## general improvements

- [x] netdata was not parsing very big double numbers properly. When numbers were way above the max value that can fit in 64-bits, the parsing was actually giving wrong values. Fixed it and also allowed faster parsing even in scientific notation. Optimizations for 32bit systems have been added too.
- [x] added internal API to parse and print integer and double numbers in decimal, hex and base64 encodings.
- [x] all printing and parsing of numbers is now inlined for best performance.
- [x] added more unit-tests to ensure printing and parsing of numbers is consistent
- [x] make sure the dbengine metrics registry does not accept negative timestamps (due to corrupted files, or other anomalies)
- [x] make sure the validation of pages loaded from disk does not accept pages with negative timestamps
- [x] make sure the page cache does not accept pages with negative timestamps

## streaming
- [x] added `IEEE754` streaming capability, to allow 2 netdata agents that use this format for double precision values, to transfer these values base64 encoded.
- [x] streaming v2 and replication use the new `IEEE754` capability to transfer all values in base64. This increases the bandwidth somewhat, but it lowers the processing power on busy parents.

## weights endpoint
- [x] weights is now doing synchronous dbengine queries - it seems they are twice as fast for it.

## JSON formatting API in `buffer_X()`
- [x] provide a json formatting API under the `buffer_json_X()` set of functions. Now generating valid json output is as easy as opening a json object and adding members to it (all types, including sub-objects, members, arrays, etc).
- [x] escape all strings, both keys and values, added in json output
- [x] optimize buffer core functions (strcat, number printing, etc)
- [x] re-write `jsonwrap` to use the new json formatting
- [x] re-write `contexts` to use the new json formatting
- [x] re-write `info` to use the new new json formatting
- [x] re-write `functions` to use the new json formatting
- [x] re-write `weights` to use the new json formatting
- [x] re-write `apps.plugin` functions to use the new json formatting

### CODE REVIEW

File: libnetdata/buffer/buffer.c [hunk 1]
```diff
@@ -585,3 +323,93 @@
...
+    buffer_print_int64_hex(wb, -1676071986);
+    errors += buffer_expect(wb, "-0x63E6D432");
+
+    buffer_flush(wb);
+
+    {
+        char buf[1024 + 1];
+        for(size_t i = 0; i < 1024 ;i++)
+            buf[i] = (char)(i % 26) + 'A';
+        buf[8192] = '\0';
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply libnetdata/buffer/buffer.c]: fixed it

**Comment by stelfrag** [comment]
Buffer test fails (and because of that unit tests)  with 

```
2023-02-15 12:12:29: netdata ERROR : MAIN : BUFFER: mismatch.
Generated:
9.123456789012346128e+45
Expected:
9.12345678901234614e+45

2023-02-15 12:12:29: netdata ERROR : MAIN : BUFFER: mismatch.
Generated:
{
}

Expected:
{
}

2023-02-15 12:12:29: netdata ERROR : MAIN : BUFFER: mismatch.
Generated:
{
    "hello":"world",
    "alpha":"this: \" is a double quote",
    "object1":{
        "hello":"world"
    }
}

Expected:
{
    "hello":"world",
    "alpha":"this: \" is a double quote",
    "object1":{
        "hello":"world"
    }
}
```

**Comment by ktsaou** [comment]
Fixed it

**Comment by ktsaou** [comment]
Ready for merge

**Review by stelfrag — APPROVED** [review]

**Pull request merged**